### ISSUE TITLE
[BUG]: Facing AttributeError while using zenml up command in Windows 11 (Local)

### ISSUE THREAD

**Issue by srini047** [zenml-io/zenml#974]
### Contact Details [Optional]

srinikethcr7@gmail.com

### System Information

ZenML version: 0.20.2
Install path: C:\Users\Sriniketh J\AppData\Local\Programs\Python\Python39\Lib\site-packages\zenml
Python version: 3.9.11
Platform information: {'os': 'windows', 'windows_version_release': '10', 'windows_version': '10.0.22000', 'windows_version_service_pack': 'SP0', 'windows_version_os_type': 'Multiprocessor Free'}
Environment: native
Integrations: ['evidently', 'github', 'mlflow', 'pillow', 'plotly', 'pytorch', 'scipy', 'sklearn', 'wandb', 'xgboost']

### What happened?

![image](https://user-images.githubusercontent.com/81156510/195895677-faeab2a1-8a64-48b9-94c2-97c11382abd7.png)

Facing this error while trying to spin up the zenml server.

### Reproduction steps

1. `zenml up`

### Relevant log output

```shell
Deploying a local ZenML server with name 'local'.
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│                                                                                                  │
│ C:\Users\Sriniketh J\AppData\Local\Programs\Python\Python39\lib\runpy.py:197 in                  │
│ _run_module_as_main                                                                              │
│                                                                                                  │
│   194 │   main_globals = sys.modules["__main__"].__dict__                                        │
│   195 │   if alter_argv:                                                                         │
│   196 │   │   sys.argv[0] = mod_spec.origin                                                      │
│ ❱ 197 │   return _run_code(code, main_globals, None,                                             │
│   198 │   │   │   │   │    "__main__", mod_spec)                                                 │
│   199                                                                                            │
│   200 def run_module(mod_name, init_globals=None,                                                │
│ C:\Users\Sriniketh J\AppData\Local\Programs\Python\Python39\lib\runpy.py:87 in _run_code         │
│                                                                                                  │
│    84 │   │   │   │   │      __loader__ = loader,                                                │
│    85 │   │   │   │   │      __package__ = pkg_name,                                             │
│    86 │   │   │   │   │      __spec__ = mod_spec)                                                │
│ ❱  87 │   exec(code, run_globals)                                                                │
│    88 │   return run_globals                                                                     │
│    89                                                                                            │
│    90 def _run_module_code(code, init_globals=None,                                              │
│                                                                                                  │
│ C:\Users\Sriniketh J\AppData\Local\Programs\Python\Python39\Scripts\zenml.exe\__main__.py:7 in   │
│ <module>                                                                                         │
│                                                                                                  │
│ [Errno 2] No such file or directory: 'C:\\Users\\Sriniketh                                       │
│ J\\AppData\\Local\\Programs\\Python\\Python39\\Scripts\\zenml.exe\\__main__.py'                  │
│                                                                                                  │
│ C:\Users\Sriniketh J\AppData\Local\Programs\Python\Python39\lib\site-packages\click\core.py:1130 │
│ in __call__                                                                                      │
│                                                                                                  │
│   1127 │                                                                                         │
│   1128 │   def __call__(self, *args: t.Any, **kwargs: t.Any) -> t.Any:                           │
│   1129 │   │   """Alias for :meth:`main`."""                                                     │
│ ❱ 1130 │   │   return self.main(*args, **kwargs)                                                 │
│   1131                                                                                           │
│   1132                                                                                           │
│   1133 class Command(BaseCommand):                                                               │
│                                                                                                  │
│ C:\Users\Sriniketh J\AppData\Local\Programs\Python\Python39\lib\site-packages\click\core.py:1055 │
│ in main                                                                                          │
│                                                                                                  │
│   1052 │   │   try:                                                                              │
│   1053 │   │   │   try:                                                                          │
│   1054 │   │   │   │   with self.make_context(prog_name, args, **extra) as ctx:                  │
│ ❱ 1055 │   │   │   │   │   rv = self.invoke(ctx)                                                 │
│   1056 │   │   │   │   │   if not standalone_mode:                                               │
│   1057 │   │   │   │   │   │   return rv                                                         │
│   1058 │   │   │   │   │   # it's not safe to `ctx.exit(rv)` here!                               │
│                                                                                                  │
│ C:\Users\Sriniketh J\AppData\Local\Programs\Python\Python39\lib\site-packages\click\core.py:1657 │
│ in invoke                                                                                        │
│                                                                                                  │
│   1654 │   │   │   │   super().invoke(ctx)                                                       │
│   1655 │   │   │   │   sub_ctx = cmd.make_context(cmd_name, args, parent=ctx)                    │
│   1656 │   │   │   │   with sub_ctx:                                                             │
│ ❱ 1657 │   │   │   │   │   return _process_result(sub_ctx.command.invoke(sub_ctx))               │
│   1658 │   │                                                                                     │
│   1659 │   │   # In chain mode we create the contexts step by step, but after the                │
│   1660 │   │   # base command has been invoked.  Because at that point we do not                 │
│                                                                                                  │
│ C:\Users\Sriniketh J\AppData\Local\Programs\Python\Python39\lib\site-packages\click\core.py:1404 │
│ in invoke                                                                                        │
│                                                                                                  │
│   1401 │   │   │   echo(style(message, fg="red"), err=True)                                      │
│   1402 │   │                                                                                     │
│   1403 │   │   if self.callback is not None:                                                     │
│ ❱ 1404 │   │   │   return ctx.invoke(self.callback, **ctx.params)                                │
│   1405 │                                                                                         │
│   1406 │   def shell_complete(self, ctx: Context, incomplete: str) -> t.List["CompletionItem"]:  │
│   1407 │   │   """Return a list of completions for the incomplete value. Looks                   │
│                                                                                                  │
│ C:\Users\Sriniketh J\AppData\Local\Programs\Python\Python39\lib\site-packages\click\core.py:760  │
│ in invoke                                                                                        │
│                                                                                                  │
│    757 │   │                                                                                     │
│    758 │   │   with augment_usage_errors(__self):                                                │
│    759 │   │   │   with ctx:                                                                     │
│ ❱  760 │   │   │   │   return __callback(*args, **kwargs)                                        │
│    761 │                                                                                         │
│    762 │   def forward(                                                                          │
│    763 │   │   __self, __cmd: "Command", *args: t.Any, **kwargs: t.Any  # noqa: B902             │
│                                                                                                  │
│ C:\Users\Sriniketh                                                                               │
│ J\AppData\Local\Programs\Python\Python39\lib\site-packages\zenml\cli\server.py:157 in up         │
│                                                                                                  │
│   154 │                                                                                          │
│   155 │   server_config = ServerDeploymentConfig(**config_attrs)                                 │
│   156 │                                                                                          │
│ ❱ 157 │   server = deployer.deploy_server(server_config)                                         │
│   158 │                                                                                          │
│   159 │   gc = GlobalConfiguration()                                                             │
│   160 │   assert gc.store is not None                                                            │
│                                                                                                  │
│ C:\Users\Sriniketh J\AppData\Local\Programs\Python\Python39\lib\site-packages\zenml\zen_server\d │
│ eploy\deployer.py:118 in deploy_server                                                           │
│                                                                                                  │
│   115 │   │   │   f"Deploying a {provider_name} ZenML server with name "                         │
│   116 │   │   │   f"'{config.name}'."                                                            │
│   117 │   │   )                                                                                  │
│ ❱ 118 │   │   return provider.deploy_server(config, timeout=timeout)                             │
│   119 │                                                                                          │
│   120 │   def update_server(                                                                     │
│   121 │   │   self,                                                                              │
│                                                                                                  │
│ C:\Users\Sriniketh J\AppData\Local\Programs\Python\Python39\lib\site-packages\zenml\zen_server\d │
│ eploy\base_provider.py:118 in deploy_server                                                      │
│                                                                                                  │
│   115 │   │   # convert the generic deployment config to a provider specific                     │
│   116 │   │   # deployment config                                                                │
│   117 │   │   config = self._convert_config(config)                                              │
│ ❱ 118 │   │   service = self._create_service(config, timeout)                                    │
│   119 │   │   return self._get_deployment(service)                                               │
│   120 │                                                                                          │
│   121 │   def update_server(                                                                     │
│                                                                                                  │
│ C:\Users\Sriniketh J\AppData\Local\Programs\Python\Python39\lib\site-packages\zenml\zen_server\d │
│ eploy\local\local_provider.py:158 in _create_service                                             │
│                                                                                                  │
│   155 │   │   │   ),                                                                             │
│   156 │   │   )                                                                                  │
│   157 │   │   service = LocalZenServer(config=service_config, endpoint=endpoint)                 │
│ ❱ 158 │   │   service.start(timeout=timeout)                                                     │
│   159 │   │   return service                                                                     │
│   160 │                                                                                          │
│   161 │   def _update_service(                                                                   │
│                                                                                                  │
│ C:\Users\Sriniketh J\AppData\Local\Programs\Python\Python39\lib\site-packages\zenml\zen_server\d │
│ eploy\local\local_zen_server.py:188 in start                                                     │
│                                                                                                  │
│   185 │   │   │   │   the service status.                                                        │
│   186 │   │   """                                                                                │
│   187 │   │   if not self.config.blocking:                                                       │
│ ❱ 188 │   │   │   super().start(timeout)                                                         │
│   189 │   │   else:                                                                              │
│   190 │   │   │   self._copy_global_configuration()                                              │
│   191 │   │   │   try:                                                                           │
│                                                                                                  │
│ C:\Users\Sriniketh J\AppData\Local\Programs\Python\Python39\lib\site-packages\zenml\services\loc │
│ al\local_service.py:420 in start                                                                 │
│                                                                                                  │
│   417 │   │   │   │   the service status.                                                        │
│   418 │   │   """                                                                                │
│   419 │   │   if not self.config.blocking:                                                       │
│ ❱ 420 │   │   │   super().start(timeout)                                                         │
│   421 │   │   else:                                                                              │
│   422 │   │   │   self.run()                                                                     │
│   423                                                                                            │
│                                                                                                  │
│ C:\Users\Sriniketh                                                                               │
│ J\AppData\Local\Programs\Python\Python39\lib\site-packages\zenml\services\service.py:389 in      │
│ start                                                                                            │
│                                                                                                  │
│   386 │   │   """                                                                                │
│   387 │   │   with console.status(f"Starting service '{self}'.\n"):                              │
│   388 │   │   │   self.admin_state = ServiceState.ACTIVE                                         │
│ ❱ 389 │   │   │   self.provision()                                                               │
│   390 │   │   │   if timeout > 0:                                                                │
│   391 │   │   │   │   if not self.poll_service_status(timeout):                                  │
│   392 │   │   │   │   │   raise RuntimeError(                                                    │
│                                                                                                  │
│ C:\Users\Sriniketh J\AppData\Local\Programs\Python\Python39\lib\site-packages\zenml\zen_server\d │
│ eploy\local\local_zen_server.py:168 in provision                                                 │
│                                                                                                  │
│   165 │   def provision(self) -> None:                                                           │
│   166 │   │   """Provision the service."""                                                       │
│   167 │   │   self._copy_global_configuration()                                                  │
│ ❱ 168 │   │   super().provision()                                                                │
│   169 │                                                                                          │
│   170 │   def deprovision(self, force: bool = False) -> None:                                    │
│   171 │   │   """Deprovision the service.                                                        │
│                                                                                                  │
│ C:\Users\Sriniketh J\AppData\Local\Programs\Python\Python39\lib\site-packages\zenml\services\loc │
│ al\local_service.py:401 in provision                                                             │
│                                                                                                  │
│   398 │                                                                                          │
│   399 │   def provision(self) -> None:                                                           │
│   400 │   │   """Provision the service."""                                                       │
│ ❱ 401 │   │   self._start_daemon()                                                               │
│   402 │                                                                                          │
│   403 │   def deprovision(self, force: bool = False) -> None:                                    │
│   404 │   │   """Deprovision the service.                                                        │
│                                                                                                  │
│ C:\Users\Sriniketh J\AppData\Local\Programs\Python\Python39\lib\site-packages\zenml\services\loc │
│ al\local_service.py:348 in _start_daemon                                                         │
│                                                                                                  │
│   345 │   │   logger.debug("Starting daemon for service '%s'...", self)                          │
│   346 │   │                                                                                      │
│   347 │   │   if self.endpoint:                                                                  │
│ ❱ 348 │   │   │   self.endpoint.prepare_for_start()                                              │
│   349 │   │                                                                                      │
│   350 │   │   command, command_env = self._get_daemon_cmd()                                      │
│   351 │   │   logger.debug(                                                                      │
│                                                                                                  │
│ C:\Users\Sriniketh J\AppData\Local\Programs\Python\Python39\lib\site-packages\zenml\services\loc │
│ al\local_service_endpoint.py:128 in prepare_for_start                                            │
│                                                                                                  │
│   125 │   │   """                                                                                │
│   126 │   │   self.status.protocol = self.config.protocol                                        │
│   127 │   │   self.status.hostname = self.config.ip_address                                      │
│ ❱ 128 │   │   self.status.port = self._lookup_free_port()                                        │
│   129                                                                                            │
│                                                                                                  │
│ C:\Users\Sriniketh J\AppData\Local\Programs\Python\Python39\lib\site-packages\zenml\services\loc │
│ al\local_service_endpoint.py:107 in _lookup_free_port                                            │
│                                                                                                  │
│   104 │   │   """                                                                                │
│   105 │   │   # If a port value is explicitly configured, attempt to use it first                │
│   106 │   │   if self.config.port:                                                               │
│ ❱ 107 │   │   │   if port_available(self.config.port, self.config.ip_address):                   │
│   108 │   │   │   │   return self.config.port                                                    │
│   109 │   │   │   if not self.config.allocate_port:                                              │
│   110 │   │   │   │   raise IOError(f"TCP port {self.config.port} is not available.")            │
│                                                                                                  │
│ C:\Users\Sriniketh                                                                               │
│ J\AppData\Local\Programs\Python\Python39\lib\site-packages\zenml\utils\networking_utils.py:40 in │
│ port_available                                                                                   │
│                                                                                                  │
│    37 │   try:                                                                                   │
│    38 │   │   with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:                       │
│    39 │   │   │   s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)                        │
│ ❱  40 │   │   │   s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)                        │
│    41 │   │   │   s.bind((address, port))                                                        │
│    42 │   except socket.error as e:                                                              │
│    43 │   │   logger.debug("Port %d unavailable on %s: %s", port, address, e)
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by srini047** [comment]
- Possible [fix](https://stackoverflow.com/questions/34207093/python-code-works-in-ubuntu-14-04-but-not-in-win8-1)
- File (line numbers) causing the issue: https://github.com/zenml-io/zenml/blob/main/src/zenml/utils/networking_utils.py#L39-L40

**Comment by dnth** [comment]
Hi @srini047 thank you for reporting it here!

**Comment by htahir1** [comment]
This should be  fixed with 0.20.5 :-) . Please reopen if you run into it again @srini047 !
